[FOR DISCUSSION] 21368-Integrate-FileAttributesPlugin

akgrant43 opened the pull request:
For discussion only.

This PR merges in all the changes required to the image to support the (not so) new FileAttributesPlugin.

Note that this PR can't be reviewed by pulling the changes in to an existing image since it will attempt to perform file access in the middle of the import.   The updated image needs to be bootstrapped.

akgrant committed:
21368 Integrate FileAttributesPlugin

Modify Files and FileSystem to use the enhanced functionality provided by FileAttributesPlugin.

akgrant committed:
Reset file mode masks on startup

akgrant committed:
21368 Initialise File in PharoBootstrapInitialization class>>initializeFileSystem

File loads a number of constants from the VM along with keeping a
character encoder.

akgrant committed:
Fixes for Windows specific code introduced by latest refactoring

akgrant committed:
Merge ebc8ea4d4f33f93e340d2801efd36745eb8c5d2c

jecisc commented on src/FileSystem-Core/DiskDirectoryEntry.class.st:
```diff
@@ -0,0 +1,338 @@
+
+{ #category : #'testing - stat' }
+DiskDirectoryEntry >> isCharacter [
+	"Answer true if the receiver is a character device"
```
What is a character device?

akgrant43 replied:
Hi Cyril,

> What is a character device?

The example I'm aware of is /dev/urandom.  But not really having a better understanding, here's a couple of pages I found:

"A character device typically transfers data to and from a user application — they behave like pipes or serial ports, instantly reading or writing the byte data in a character-by-character stream. They provide the framework for many typical drivers, such as those that are required for interfacing to serial communications, video capture, and audio devices. "

- http://derekmolloy.ie/writing-a-linux-kernel-module-part-2-a-character-device/

https://unix.stackexchange.com/questions/37829/how-do-character-device-or-character-special-files-work also has quite a bit of information, e.g. that character devices don't support the normal seek and tell operations.

akgrant committed:
The VM will return Squeak times instead of Unix times

akgrant committed:
File attributes #uid, #gid and #deviceId

Lost during latest refactor

akgrant committed:
Merge remote-tracking branch 'origin/development' into 21368-mergedev

akgrant committed:
File attribute #inode

Lost during latest refactor

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

pavel-krivanek commented:
the PR produces two new Undeclared (#WindowsFileAttributesPluginPrims #FileAttributePrimitives) and introduces new bad dependencies

akgrant43 commented:
Hi @pavel-krivanek ,

Thanks very much for your reply.

The two undeclared variables were in a no-longer-used method that
somehow got left behind.  I've removed it but will hold off pushing the
commit until I've addressed the bad dependencies.

I've also resolved the conflicts in FileReferenceTest (also still to be
pushed up).

Would you mind clarifying what the bad dependencies are?


Thanks again,
Alistair

akgrant committed:
21368 - Remove WindowsStore class>>useFilePlugin

This is a leftover from before the FilePluginPrims refactoring.

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant43 commented:
A few comments on the CI tests, which reported 29 failures:

- 5 of the failures are reported twice - all the "New failing" are also "Existing failing".  Is that normal?
- I couldn't reproduce the Unix 32 ZipArchiveTest failures.
- The 14 DiskFileAttributesTests and FileReferenceAttributeTests failures are expected.  The changes require a VM dated 17 Jun or later, and the new VM was promoted just after the build started.  They should be green in the next run.

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant43 commented:
Some real issues to be addressed in the latest build...

akgrant committed:
21368: DiskFileAttributesTests>>setUp call super setUp

akgrant committed:
21368: Add MemoryStore missing attributes

#isReadable:, #isRegular:, #isWritable, #permissions:

akgrant committed:
Merge 75a18806c7e04577c8650978ef3e11481235f311

akgrant committed:
21368 - Refactor to undo adding FileSystem dependency to Files

akgrant committed:
21368 - redo DiskFileAttributesTests>>setUp

#setUp should call super setUp as the first message.

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-Integrate-FileAttributesPlugin

akgrant committed:
21368 additional descriptions for FileSystemTest>>testFileNames

akgrant committed:
Merge 5a73631027bc93c889f6d3e30b9115363fb6b397

akgrant committed:
Merge remote-tracking branch 'origin/21368-mergedev' into 21368-Integrate-FileAttributesPlugin

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
Merge branch '21368-mergedev' into 21368-Integrate-FileAttributesPlugin

akgrant committed:
21368: FileSystemTest>>testFileNames additional debug information

akgrant committed:
Merge 1844fe99bc06441a6cc07823d1f279020332c43d

akgrant committed:
Merge remote-tracking branch 'origin/21368-mergedev' into 21368-Integrate-FileAttributesPlugin

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
FileAttributesPlugin: Fix DiskStore>>isSymlink (was still using FilePlugin), Remove platform hack in File class>>isSymlink:

akgrant committed:
Merge remote-tracking branch 'origin/21368-mergedev' into 21368-mergedev

akgrant committed:
Merge remote-tracking branch 'upstream/development' into 21368-mergedev

akgrant committed:
FileAttributes: DiskDirectoryEntry>>isSymlink failed on Windows

Delegate the test to FileReference so that platform specific processing is performed.

akgrant committed:
FileAttributes: VM now flags when symlinks aren't supported

akgrant committed:
Merge remote-tracking branch 'origin/21368-mergedev' into 21368-Integrate-FileAttributesPlugin

estebanlm commented:
ok, please reopen if still relevant :)

akgrant43 commented:
Hi Esteban,

Shall do.

I've since learnt more than I ever wanted to know about file name encoding on OSX.  I've also, I think, battled my way through file name encoding on Windows - functions that only accept ascii, functions that pretend to accept UTF8 but don't (I hate Windows), functions that require the "\\?\" prefix for long file names, functions that accept the "\\?\" prefix but then fail depending on the file name pattern (did I mention that I hate Windows?), ...

Anyway, I think I've got the VM working properly now (still to be committed), and am tidying up the smalltalk code.  After that I should retest all three platforms.  And then I can reopen the PR!

Cheers,
Alistair

bencoman commented:
Hi Alistair,

Thanks for your effort and persistence with this.  I appreciate not needing
to personally plumb such depths of file name encoding if you can
successfully encapsulate it.

cheers -ben


On Fri, 28 Sep 2018 at 00:40, akgrant43 <notifications@github.com> wrote:

> Hi Esteban,
>
> Shall do.
>
> I've since learnt more than I ever wanted to know about file name encoding
> on OSX. I've also, I think, battled my way through file name encoding on
> Windows - functions that only accept ascii, functions that pretend to
> accept UTF8 but don't (I hate Windows), functions that require the "\?"
> prefix for long file names, functions that accept the "\?" prefix but then
> fail depending on the file name pattern (did I mention that I hate
> Windows?), ...
>
> Anyway, I think I've got the VM working properly now (still to be
> committed), and am tidying up the smalltalk code. After that I should
> retest all three platforms. And then I can reopen the PR!
>
> Cheers,
> Alistair
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pharo-project/pharo/pull/1529#issuecomment-425161172>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABolJ8IgcMpTU9DfJqgVYxWdr2JnGG3Kks5ufP9ogaJpZM4UkR81>
> .
>

akgrant43 commented:
Hi Ben,

Thanks for your encouragement.  I have to admit that I'll be happy when I can spend more time programming in Smalltalk! :-)

Cheers,
Alistair

akgrant43 commented:
This PR has been superseeded by #2431
